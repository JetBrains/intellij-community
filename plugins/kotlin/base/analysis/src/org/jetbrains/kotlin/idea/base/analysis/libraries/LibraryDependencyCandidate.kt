// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.base.analysis.libraries

import org.jetbrains.annotations.ApiStatus
import org.jetbrains.kotlin.idea.base.projectStructure.moduleInfo.AbstractKlibLibraryInfo
import org.jetbrains.kotlin.idea.base.projectStructure.moduleInfo.LibraryInfo
import org.jetbrains.kotlin.platform.TargetPlatform

@ApiStatus.Internal
sealed class LibraryDependencyCandidate {
    abstract val platform: TargetPlatform
    abstract val libraries: List<LibraryInfo>

    companion object {
        fun fromLibraryOrNull(libraryInfos: List<LibraryInfo>): LibraryDependencyCandidate? {
            val libraryInfo = libraryInfos.firstOrNull() ?: return null
            return if (libraryInfo is AbstractKlibLibraryInfo) {
                KlibLibraryDependencyCandidate(
                    platform = libraryInfo.platform,
                    libraries = libraryInfos,
                    uniqueName = libraryInfo.uniqueName,
                    isInterop = libraryInfo.isInterop
                )
            } else {
                DefaultLibraryDependencyCandidate(
                    platform = libraryInfo.platform,
                    libraries = libraryInfos
                )
            }
        }
    }
}

@ApiStatus.Internal
data class DefaultLibraryDependencyCandidate(
    override val platform: TargetPlatform,
    override val libraries: List<LibraryInfo>
): LibraryDependencyCandidate()

@ApiStatus.Internal
data class KlibLibraryDependencyCandidate(
  override val platform: TargetPlatform,
  override val libraries: List<LibraryInfo>,
  val uniqueName: String?,
  val isInterop: Boolean
): LibraryDependencyCandidate()