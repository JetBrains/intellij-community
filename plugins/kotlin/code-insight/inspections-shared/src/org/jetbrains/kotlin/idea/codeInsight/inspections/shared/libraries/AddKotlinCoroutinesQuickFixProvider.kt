// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries

import com.intellij.openapi.module.Module
import com.intellij.openapi.module.impl.scopes.ModulesScope
import com.intellij.psi.JavaPsiFacade

internal class AddKotlinCoroutinesQuickFixProvider : SimpleAddKotlinLibraryQuickFixProvider(
    libraryGroupId = "org.jetbrains.kotlinx",
    libraryArtifactId = "kotlinx-coroutines-core",
    namesToCheck = setOf("runBlocking", "CoroutineScope")
) {
    override fun hasLibrary(module: Module): Boolean {
        val scope = ModulesScope.moduleWithDependenciesAndLibrariesScope(module)
        return JavaPsiFacade.getInstance(module.project).findClasses("kotlinx.coroutines.CoroutineScope", scope).isNotEmpty()
    }
}