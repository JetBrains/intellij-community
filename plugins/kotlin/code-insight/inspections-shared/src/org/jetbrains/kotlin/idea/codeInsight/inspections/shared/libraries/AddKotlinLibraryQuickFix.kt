// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries

import com.intellij.codeInsight.intention.HighPriorityAction
import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.codeInspection.util.IntentionName
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ExternalLibraryDescriptor
import com.intellij.psi.PsiFile
import org.jetbrains.kotlin.idea.base.util.module
import org.jetbrains.kotlin.idea.configuration.KotlinBuildSystemDependencyManager
import org.jetbrains.kotlin.idea.configuration.isProjectSyncPendingOrInProgress

internal class AddKotlinLibraryQuickFix(
    private val dependencyManager: KotlinBuildSystemDependencyManager,
    private val libraryDescriptor: ExternalLibraryDescriptor,
    @IntentionName
    private val quickFixText: String
) : IntentionAction, HighPriorityAction {
    override fun getText(): String = quickFixText
    override fun getFamilyName(): String = quickFixText
    override fun startInWriteAction(): Boolean = false
    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
        val module = file?.module ?: return false
        return dependencyManager.isApplicable(module) && !dependencyManager.isProjectSyncPendingOrInProgress()
    }

    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
        if (editor == null && !ApplicationManager.getApplication().isHeadlessEnvironment) return
        if (file == null) return
        val psiFile = file.originalFile
        val module = psiFile.module ?: return

        ApplicationManager.getApplication().runWriteAction {
            dependencyManager.addDependency(module, libraryDescriptor)
        }

        dependencyManager.startProjectSync()

        val buildScriptFile = dependencyManager.getBuildScriptFile(module)
        if (buildScriptFile != null) {
            FileEditorManager.getInstance(module.project).openFile(buildScriptFile, false)
        }
    }
}