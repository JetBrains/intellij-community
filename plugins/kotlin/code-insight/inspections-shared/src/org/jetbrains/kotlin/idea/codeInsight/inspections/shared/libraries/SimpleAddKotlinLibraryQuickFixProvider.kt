// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries

import com.intellij.psi.PsiReference
import org.jetbrains.kotlin.psi.KtQualifiedExpression
import org.jetbrains.kotlin.psi.KtReferenceExpression

/**
 * A simple [AddKotlinLibraryQuickFixProvider] that determines if an unresolved reference belongs to the library if they
 * are contained in the [namesToCheck].
 * The names that are checked are references or function calls that are not part of a dot qualified expression.
 */
abstract class SimpleAddKotlinLibraryQuickFixProvider(
    libraryGroupId: String,
    libraryArtifactId: String,
    private val namesToCheck: Set<String>,
) : AddKotlinLibraryQuickFixProvider(libraryGroupId, libraryArtifactId) {
    override fun isLibraryReference(ref: PsiReference): Boolean {
        val referenceExpression = ref.element as? KtReferenceExpression ?: return false
        if (!namesToCheck.contains(referenceExpression.text)) return false
        return referenceExpression.parent !is KtQualifiedExpression
    }
}