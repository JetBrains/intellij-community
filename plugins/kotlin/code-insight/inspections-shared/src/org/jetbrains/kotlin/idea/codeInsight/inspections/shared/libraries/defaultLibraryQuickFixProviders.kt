// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries

import org.jetbrains.kotlin.idea.base.resources.KotlinBundle

internal class AddKotlinCoroutinesQuickFixProvider : AddKotlinLibraryQuickFixProvider(
    libraryGroupId = "org.jetbrains.kotlinx",
    libraryArtifactId = "kotlinx-coroutines-core",
    libraryDescriptorProvider = LibraryDescriptorProvider.default(),
    libraryAvailabilityTester = LibraryAvailabilityTester.knownClassFqn("kotlinx.coroutines.CoroutineScope"),
    libraryReferenceTester = LibraryReferenceTester.knownNames(
        "runBlocking", "CoroutineScope", "Dispatchers", "launch", "GlobalScope",
        "MainScope", "async", "await", "withContext", "cancel", "isActive", "Job",
    ),
    quickFixText = KotlinBundle.message("add.kotlin.coroutines")
)


internal class AddKotlinTestLibraryQuickFixProvider : AddKotlinLibraryQuickFixProvider(
    libraryGroupId = "org.jetbrains.kotlin",
    libraryArtifactId = "kotlin-test",
    libraryDescriptorProvider = LibraryDescriptorProvider.default(),
    libraryAvailabilityTester = LibraryAvailabilityTester.knownClassFqn("kotlin.test.Asserter"),
    libraryReferenceTester = LibraryReferenceTester.knownNames("Test", "assertEquals", "assertTrue", "assertFalse", "assertFailsWith"),
)
