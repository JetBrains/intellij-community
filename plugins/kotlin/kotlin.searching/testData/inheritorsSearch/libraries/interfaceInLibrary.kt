// WITH_LIBRARY: libraryInterfaceInLibraryB.kt
fun main() {}