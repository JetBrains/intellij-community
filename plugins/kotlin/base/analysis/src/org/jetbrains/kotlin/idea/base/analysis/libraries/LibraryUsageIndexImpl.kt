// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.base.analysis.libraries

import com.intellij.java.library.JavaLibraryModificationTracker
import com.intellij.openapi.application.runReadAction
import com.intellij.openapi.module.ModuleManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.module.Module
import com.intellij.openapi.roots.LibraryOrderEntry
import com.intellij.openapi.roots.ModuleRootManager
import com.intellij.openapi.roots.libraries.Library
import com.intellij.psi.util.CachedValue
import com.intellij.psi.util.CachedValueProvider
import org.jetbrains.kotlin.idea.base.facet.isHMPPEnabled
import com.intellij.psi.util.CachedValuesManager
import com.intellij.util.containers.MultiMap
import org.jetbrains.kotlin.idea.base.projectStructure.LibraryInfoCache
import org.jetbrains.kotlin.idea.base.projectStructure.LibraryUsageIndex
import org.jetbrains.kotlin.idea.base.projectStructure.moduleInfo.LibraryInfo
import org.jetbrains.kotlin.idea.caches.project.canDependOn
import org.jetbrains.kotlin.idea.caches.project.getIdeaModelInfosCache
import org.jetbrains.kotlin.idea.caches.trackers.ModuleModificationTracker
import org.jetbrains.kotlin.progress.ProgressIndicatorAndCompilationCanceledStatus.checkCanceled

class LibraryUsageIndexImpl(private val project: Project) : LibraryUsageIndex {
    private val moduleDependentsByLibrary: CachedValue<MultiMap<Library, Module>> =
        CachedValuesManager.getManager(project).createCachedValue {
            CachedValueProvider.Result(
                computeLibraryModuleDependents(),
                ModuleModificationTracker.getInstance(project),
                JavaLibraryModificationTracker.getInstance(project),
            )
        }

    override fun getDependentModules(libraryInfo: LibraryInfo): Sequence<Module> = sequence<Module> {
        val ideaModelInfosCache = getIdeaModelInfosCache(project)
        for (module in moduleDependentsByLibrary.value[libraryInfo.library]) {
            val mappedModuleInfos = ideaModelInfosCache.getModuleInfosForModule(module)
            if (mappedModuleInfos.any { it.platform.canDependOn(libraryInfo, module.isHMPPEnabled) }) {
                yield(module)
            }
        }
    }

    private fun computeLibraryModuleDependents(): MultiMap<Library, Module> = runReadAction {
        val moduleDependentsByLibrary = MultiMap.createSet<Library, Module>()
        val libraryCache = LibraryInfoCache.getInstance(project)
        for (module in ModuleManager.getInstance(project).modules) {
            checkCanceled()
            for (entry in ModuleRootManager.getInstance(module).orderEntries) {
                if (entry !is LibraryOrderEntry) continue
                val library = entry.library ?: continue
                val keyLibrary = libraryCache.deduplicatedLibrary(library)
                moduleDependentsByLibrary.putValue(keyLibrary, module)
            }
        }

        moduleDependentsByLibrary
    }
}
