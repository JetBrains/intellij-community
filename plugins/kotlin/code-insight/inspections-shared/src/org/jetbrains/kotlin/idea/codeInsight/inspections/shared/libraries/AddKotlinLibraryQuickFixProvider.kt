// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries

import com.intellij.codeInsight.daemon.QuickFixActionRegistrar
import com.intellij.codeInsight.quickfix.UnresolvedReferenceQuickFixProvider
import com.intellij.codeInspection.util.IntentionName
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.impl.scopes.ModulesScope
import com.intellij.openapi.roots.DependencyScope
import com.intellij.openapi.roots.ExternalLibraryDescriptor
import com.intellij.openapi.roots.ProjectFileIndex
import com.intellij.psi.JavaPsiFacade
import com.intellij.psi.PsiReference
import org.jetbrains.kotlin.idea.base.resources.KotlinBundle
import org.jetbrains.kotlin.idea.base.util.module
import org.jetbrains.kotlin.idea.codeInsight.inspections.shared.libraries.AddKotlinLibraryQuickFixProvider.*
import org.jetbrains.kotlin.idea.configuration.KotlinBuildSystemDependencyManager
import org.jetbrains.kotlin.idea.configuration.KotlinLibraryVersionProvider
import org.jetbrains.kotlin.idea.configuration.isProjectSyncPendingOrInProgress
import org.jetbrains.kotlin.idea.configuration.withScope
import org.jetbrains.kotlin.idea.stubindex.KotlinFullClassNameIndex
import org.jetbrains.kotlin.psi.KtQualifiedExpression
import org.jetbrains.kotlin.psi.KtReferenceExpression

/**
 * A quick fix provider allowing for Kotlin libraries to be added when an unresolved element was detected.
 * [libraryGroupId] and [libraryArtifactId] are the maven coordinates that will be added by this quick-fix.
 */
open class AddKotlinLibraryQuickFixProvider(
    private val libraryGroupId: String,
    private val libraryArtifactId: String,
    private val libraryDescriptorProvider: LibraryDescriptorProvider,
    private val libraryAvailabilityTester: LibraryAvailabilityTester,
    private val libraryReferenceTester: LibraryReferenceTester,
    @IntentionName
    private val quickFixText: String = KotlinBundle.message("add.0.library", "$libraryGroupId:$libraryArtifactId")
) : UnresolvedReferenceQuickFixProvider<PsiReference>() {

    fun interface LibraryAvailabilityTester {
        /**
         * Returns whether the library that will be added by this quick-fix is already present in the module.
         */
        fun isAvailable(module: Module): Boolean

        companion object
    }

    fun interface LibraryReferenceTester {
        /**
         * Checks if the reference is a reference to a definition of the library belonging to this quick-fix.
         */
        fun isLibraryReference(ref: PsiReference): Boolean

        companion object
    }

    fun interface LibraryDescriptorProvider {
        fun getLibraryDescriptor(libraryGroupId: String, libraryArtifactId: String, psiReference: PsiReference): ExternalLibraryDescriptor?

        companion object
    }

    override fun registerFixes(ref: PsiReference, registrar: QuickFixActionRegistrar) {
        if (!libraryReferenceTester.isLibraryReference(ref)) return
        val module = ref.element.module ?: return
        val extensionList = ref.element.project.extensionArea.getExtensionPoint(KotlinBuildSystemDependencyManager.EP_NAME).extensionList
        val dependencyManager = extensionList.firstOrNull { it.isApplicable(module) } ?: return
        if (dependencyManager.isProjectSyncPendingOrInProgress()) return

        if (libraryAvailabilityTester.isAvailable(module)) return

        val libraryDescriptor = libraryDescriptorProvider.getLibraryDescriptor(
            libraryGroupId, libraryArtifactId, ref
        ) ?: return

        val scope = if (ProjectFileIndex.getInstance(module.project).isInTestSourceContent(ref.element.containingFile.virtualFile)) {
            DependencyScope.TEST
        } else {
            DependencyScope.COMPILE
        }

        registrar.register(
            AddKotlinLibraryQuickFix(
                dependencyManager = dependencyManager,
                libraryDescriptor = libraryDescriptor.withScope(scope),
                quickFixText = quickFixText
            )
        )
    }

    override fun getReferenceClass(): Class<PsiReference> = PsiReference::class.java
}

/**
 * Will use the [KotlinLibraryVersionProvider] to look up the version
 */
fun LibraryDescriptorProvider.Companion.default() = LibraryDescriptorProvider { libraryGroupId, libraryArtifactId, psiRef ->
    val module = psiRef.element.module ?: return@LibraryDescriptorProvider null
    val libraryVersion = KotlinLibraryVersionProvider.EP_NAME.extensionList.firstNotNullOfOrNull { provider ->
        provider.getVersion(module, libraryGroupId, libraryArtifactId)
    }

    ExternalLibraryDescriptor(libraryGroupId, libraryArtifactId, libraryVersion, libraryVersion, libraryVersion)
}


/**
 * A simple [LibraryReferenceTester] that determines if an unresolved reference belongs to the library if they
 * are contained in the [names].
 * The names that are checked are references or function calls that are not part of a dot qualified expression.
 */
fun LibraryReferenceTester.Companion.knownNames(vararg names: String) = LibraryReferenceTester { ref ->
    val referenceExpression = ref.element as? KtReferenceExpression ?: return@LibraryReferenceTester false
    if (referenceExpression.parent is KtQualifiedExpression) return@LibraryReferenceTester false
    referenceExpression.text in names
}

/**
 * Will check if the provided [fqn] is available as class in the given module
 */
fun LibraryAvailabilityTester.Companion.knownClassFqn(fqn: String) = LibraryAvailabilityTester { module ->
    val scope = ModulesScope.moduleWithDependenciesAndLibrariesScope(module)
    JavaPsiFacade.getInstance(module.project).findClasses(fqn, scope).isNotEmpty() ||
            KotlinFullClassNameIndex[fqn, module.project, scope].isNotEmpty()
}