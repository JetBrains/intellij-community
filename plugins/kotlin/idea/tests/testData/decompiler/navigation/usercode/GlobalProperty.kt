import testData.libraries.*

fun foo() {
    println(testData.libraries.globalVal)
    println(globalValWithGetter)
}
