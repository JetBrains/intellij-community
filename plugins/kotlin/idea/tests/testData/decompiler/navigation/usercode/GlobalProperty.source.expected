main.kt
public val <1>globalVal : Pair<Int, String> = Pair(239, "239")

public val <2>globalValWithGetter : Long
